numpy/random/_generator.pyx

```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
         pix = <double*>np.PyArray_DATA(parr)
         check_array_constraint(parr, 'pvals', CONS_BOUNDED_0_1)
         if kahan_sum(pix, d-1) > (1.0 + 1e-12):
-            raise ValueError("sum(pvals[:-1]) > 1.0")
+            # When floating, but not float dtype, and close, improve the error
+            # 1.0001 works for float16 and float32
+            if (isinstance(pvals, np.ndarray) and
+                    pvals.dtype != float and
+                    np.issubdtype(pvals.dtype, np.floating) and
+                    pvals.sum() < 1.0001):
+                msg = ("sum(pvals[:-1].astype(np.float64)) > 1.0. pvals are "
+                       "cast to 64-bit floating point values prior to "
+                       "checking the constraint. Changes in precision when "
```

Comment:
"constraint" -> "sum"
"Changes in precision" -> "Precision changes"
"produce violations" -> "cause problems"

Reply:
Fixed.